### ISSUE TITLE
Self-Attention

### ISSUE THREAD

**Issue by ilyak93** [VITA-Group/TransGAN#14]
Why is in the self-attention layer you have a linear projection after multiplying the value matrix with the attention weights ? 
In this overview you can see, as much as I can tell that it doesn't appear: https://arxiv.org/pdf/1906.01529.pdf. 
In some implementations it doesn't appear and in some it does, what is the impact of it to the attention ?
Is it add weights to the attention for its better configuration ? 

> self.proj = nn.Linear(dim, dim)

in theAttention model:
```
class Attention(nn.Module):
    def __init__(self, dim, num_heads=8, qkv_bias=False, qk_scale=None, attn_drop=0., proj_drop=0., is_mask=0):
        super().__init__()
        self.num_heads = num_heads
        head_dim = dim // num_heads
        # NOTE scale factor was wrong in my original version, can set manually to be compat with prev weights
        self.scale = qk_scale or head_dim ** -0.5

        self.qkv = nn.Linear(dim, dim * 3, bias=qkv_bias)
        self.attn_drop = nn.Dropout(attn_drop)
        self.proj = nn.Linear(dim, dim)
        self.proj_drop = nn.Dropout(proj_drop)
        self.mat = matmul()
        self.is_mask = is_mask
        self.remove_mask = False
        self.mask_4 = get_attn_mask(is_mask, 4)
        self.mask_5 = get_attn_mask(is_mask, 5)
        self.mask_6 = get_attn_mask(is_mask, 6)
        self.mask_7 = get_attn_mask(is_mask, 7)
        self.mask_8 = get_attn_mask(is_mask, 8)
        self.mask_10 = get_attn_mask(is_mask, 10)
```

**Comment by yueruchen** [comment]
This implementation follows the ViT work (https://arxiv.org/abs/2010.11929), you can try with different variants.
